CIPs/0009.md

```diff
@@ -0,0 +1,86 @@
+
+The proposed solution is to use `clef`, introduced in [go-ethereum v1.9](https://blog.ethereum.org/2019/07/10/geth-v1-9-0/), as a standalone signer for validators.
+
+`clef` is integrated into `celo-blockchain` as an implementation of the [wallet](https://github.com/celo-org/celo-blockchain/blob/3caca596d1ff23a9faf5339f108c993f38bfa743/accounts/accounts.go#L50) interface in [accounts/external/backend.go](https://github.com/celo-org/celo-blockchain/blob/master/accounts/external/backend.go).
+
+Validators will run a `clef` instance that manages their [validator signer key](https://docs.celo.org/validator-guide/summary/detailed#authorized-validator-signers), and point their node to that instance.
+
+To switch nodes, validators will issue a command to `clef` to stop accepting RPCs from node A and start accepting RPCs from node B.
+
+### Implementation
+
+#### Wallet interface
+The current `clef` implementation of the `wallet` interface is missing the following functions needed to support running a validator:
+- Decrypt
+- SignHashBLS
+- SignMessageBLS
+- GenerateProofOfPossession
+- GenerateProofOfPossessionBLS
+- GetPublicKey
+
+To implement these functions, new RPCs will need to be added to `clef`.
+
+#### Node mutex
+More investigation will need to be done to figure out exactly how to atomically switch nodes.
+
+Some options:
+
+- Implement a `clef` RPC to switch the value of the `--rpcaddr` flag.
+- Have `clef` accept RPCs from both nodes, and use `miner.start` and `miner.stop` on the node (or similar RPCs that allow these operations to be scheduled)
+
+#### Threshold signatures
+Using threshold signatures for block finalization is an approach taken by [Polychain Labs' Cosmos validator](https://blog.polychainlabs.com/tendermint/2020/03/26/threshold-validator-for-tendermint.html) to improve security and availability.
+
+Celo is a natural candidate to adopt a similar feature for BLS block finalization signatures and ECDSA consensus message signatures.
+
+A standalone signing service would simplify the implementation of this feature.
+
+Validators could run `N` instances of `clef`, and one or more instances of `celo-blockchain`.
+
+The `celo-blockchain` nodes could be configured to ping each of the `clef` instances for a signature share, and return once `M` have been collected.
+
+The `M/N` threshold signature would protect against `N-M` `clef` instances being unavailable or compromised.
+
+Multiple `celo-blockchain` nodes with a failover mechanism would protect against a single node being unavailable.
+
+## Risks
+
+- The new node will not have the same consensus DB, which introduces a small risk of double signing. One option to mitigate this is for `clef` to stop accepting RPCs from node A and wait a block before accepting RPs from node B.
+- Upgrades to `clef` may still require key rotations. This could be mitigated by threshold signatures; so long as `M > N/2` individual `clef` instances could be upgraded one at a time with zero downtime.
```

Comment:
One advantage of this approach worth pointing out though is that `clef` would likely restart quite quickly, well under 1 minutes. `celo-blockchain` currently takes over 1 minute to fully reboot and get connected again, which is one of the main issues that is making it hard to operate with 0 downtime.